### ISSUE TITLE
Kombu disconnects on publish

### ISSUE THREAD

**Issue by stuartspotlight** [celery/kombu#857]
I've posted this elsewhere:-

https://github.com/celery/kombu/issues/463

but I figured it might be considered a separate issue so I created this issue as well.

When I do a lot of high frequency consume and publishes kombu is disconnecting on a publish command with the following error:-

```
  W_FORCE_CONNECT.format(attr=attr)))
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/nanowire_plugin/__init__.py", line 447, in send_to_next_plugin
    'max_retries':10})
  File "/usr/local/lib/python3.6/site-packages/kombu/messaging.py", line 181, in publish
    exchange_name, declare,
  File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 494, in _ensured
    return fun(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/kombu/messaging.py", line 203, in _publish
    mandatory=mandatory, immediate=immediate,
  File "/usr/local/lib/python3.6/site-packages/amqp/channel.py", line 1745, in basic_publish_confirm
    self.wait(spec.Basic.Ack)
  File "/usr/local/lib/python3.6/site-packages/amqp/abstract_channel.py", line 79, in wait
    self.connection.drain_events(timeout=timeout)
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 471, in drain_events
    while not self.blocking_read(timeout):
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 477, in blocking_read
    return self.on_inbound_frame(frame)
  File "/usr/local/lib/python3.6/site-packages/amqp/method_framing.py", line 55, in on_frame
    callback(channel, method_sig, buf, None)
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 480, in on_inbound_method
    return self.channels[channel_id].dispatch_method(
TypeError: 'NoneType' object is not subscriptable
```
This is failing to reconnect even after attempting to reconnect 10 times:-
```
            producer.publish(send_payload, exchange='', routing_key=next_plugin, retry=True, 
                             retry_policy={'interval_start':1,
                                           'interval_step':2,
                                           'interval_max':10,
                                           'max_retries':10})
```
Also worth noting is that I get the following warning when kombu attempts to reconnect:-
```
/usr/local/lib/python3.6/site-packages/amqp/connection.py:292: AMQPDeprecationWarning: The .transport attribute on the connection was accessed before
the connection was established.  This is supported for now, but will
be deprecated in amqp 2.2.0.

Since amqp 2.0 you have to explicitly call Connection.connect()
before using the connection.
```

I assume this warning will disappear when we move to kombu 4.2 but is there any word on the publish disconnect error?

**Comment by stuartspotlight** [comment]
Is this related to this issue:- https://github.com/celery/kombu/issues/795
Will it be fixed by a move to kombu 4.2?

**Comment by stuartspotlight** [comment]
Does anyone have any insight with regards to this issue since it renders the entire library borderline useless

**Comment by stuartspotlight** [comment]
More information on this bug. This seems to occur much more regularly when the time between consuming and publishing is large (more than 10 seconds) and never on the first publish once the connection is established. A fairly reliable way to generate this error is to have 2 messages which will take longer than 10 seconds to process and it will almost always trigger on the second process.

**Comment by stuartspotlight** [comment]
I have found a work around for this problem. I placed the publish command in a for loop with a try catch to repeat the publish command if it fails. This seems like a replication of the retry command however. Does anyone have any further insight into this?

**Comment by maingoh** [comment]
I am having a similar issue with the code below and kombu 4.1.0: 

```python
from kombu import Connection, Producer, Queue
import os
import time
import logging

logging.basicConfig(level='INFO')


def get_queue(channel, queue_name):
    queue = Queue(channel=channel,
                  name=queue_name,
                  routing_key=queue_name,
                  exchange=None,
                  passive=False,
                  durable=True,
                  exclusive=False,
                  auto_delete=False)
    queue.declare()
    return queue


def errback(exc, interval):
    logging.error('Error: %r', exc)
    logging.info('Retry in %s seconds.', interval)


connection = Connection(os.getenv('AMQP_URL'), transport='pyamqp', heartbeat=30, transport_options={'confirm_publish': True})

queue = get_queue(connection.default_channel, 'foo')

producer = Producer(connection)

logging.info("Sleeping")
time.sleep(130)

retry_policy = {'interval_start': 0, 'interval_step': 1, 'max_retries': 3}

while True:

    logging.info('Sending to queue %s' % queue)
    send = connection.ensure(producer, producer.publish, errback=errback)
    send({"hey": "you"}, routing_key=queue.name, retry=True, retry_policy=retry_policy)

    logging.info('Getting from queue %s' % queue)
    recv = connection.ensure(queue, queue.get, errback=errback)
    recv(no_ack=True)
    time.sleep(1)
```

Each time the sending is done it triggers a disconnection and it seems to happen only after some inactivity on the connection (simulated with the `sleep(130)`), though it should reconnect once and shouldn't trigger error in loop. When getting a message from the queue, it seems to be unbound to any channel (while it is bound to channel 1 when sending the message).

Here are my logs (I use a docker-compose app to test, I can provide it if you want)

```
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:44:19 ===
rabbitmq_1  | accepting AMQP connection <0.558.0> (172.20.0.3:44638 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:44:19 ===
rabbitmq_1  | connection <0.558.0> (172.20.0.3:44638 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Sleeping
rabbitmq_1  | 
rabbitmq_1  | =ERROR REPORT==== 5-Jun-2018::16:45:49 ===
rabbitmq_1  | closing AMQP connection <0.558.0> (172.20.0.3:44638 -> 172.20.0.2:5672):
rabbitmq_1  | missed heartbeats from client, timeout: 30s
python_client | INFO:root:Sending to queue <Queue foo -> <Exchange u''(direct) bound to chan:1> -> foo bound to chan:1>
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:29 ===
rabbitmq_1  | accepting AMQP connection <0.582.0> (172.20.0.3:44656 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:29 ===
rabbitmq_1  | connection <0.582.0> (172.20.0.3:44656 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Getting from queue <Queue foo -> <Exchange u''(direct) bound to chan:1> -> foo bound to chan:1>
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:29 ===
rabbitmq_1  | closing AMQP connection <0.582.0> (172.20.0.3:44656 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
python_client | ERROR:root:Error: error(32, 'Broken pipe')
python_client | INFO:root:Retry in 0 seconds.
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:29 ===
rabbitmq_1  | accepting AMQP connection <0.595.0> (172.20.0.3:44658 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:29 ===
rabbitmq_1  | connection <0.595.0> (172.20.0.3:44658 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Sending to queue <Queue foo -> <Exchange u''(direct) bound to chan:1> -> foo bound to chan:1>
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | closing AMQP connection <0.595.0> (172.20.0.3:44658 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | accepting AMQP connection <0.608.0> (172.20.0.3:44660 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | connection <0.608.0> (172.20.0.3:44660 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Getting from queue <Queue foo -> <Exchange u''(direct) bound to chan:None> -> foo bound to chan:None>
python_client | ERROR:root:Error: RecoverableConnectionError(None, u'connection already closed', None, u'')
python_client | INFO:root:Retry in 0 seconds.
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | closing AMQP connection <0.608.0> (172.20.0.3:44660 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | accepting AMQP connection <0.621.0> (172.20.0.3:44662 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:30 ===
rabbitmq_1  | connection <0.621.0> (172.20.0.3:44662 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Sending to queue <Queue foo -> <Exchange u''(direct) bound to chan:1> -> foo bound to chan:1>
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | closing AMQP connection <0.621.0> (172.20.0.3:44662 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | accepting AMQP connection <0.634.0> (172.20.0.3:44664 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | connection <0.634.0> (172.20.0.3:44664 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Getting from queue <Queue foo -> <Exchange u''(direct) bound to chan:None> -> foo bound to chan:None>
python_client | ERROR:root:Error: RecoverableConnectionError(None, u'connection already closed', None, u'')
python_client | INFO:root:Retry in 0 seconds.
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | closing AMQP connection <0.634.0> (172.20.0.3:44664 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | accepting AMQP connection <0.647.0> (172.20.0.3:44666 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:31 ===
rabbitmq_1  | connection <0.647.0> (172.20.0.3:44666 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Sending to queue <Queue foo -> <Exchange u''(direct) bound to chan:1> -> foo bound to chan:1>
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | closing AMQP connection <0.647.0> (172.20.0.3:44666 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | accepting AMQP connection <0.660.0> (172.20.0.3:44668 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | connection <0.660.0> (172.20.0.3:44668 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
python_client | INFO:root:Getting from queue <Queue foo -> <Exchange u''(direct) bound to chan:None> -> foo bound to chan:None>
python_client | ERROR:root:Error: RecoverableConnectionError(None, u'connection already closed', None, u'')
python_client | INFO:root:Retry in 0 seconds.
rabbitmq_1  | 
rabbitmq_1  | =WARNING REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | closing AMQP connection <0.660.0> (172.20.0.3:44668 -> 172.20.0.2:5672, vhost: 'test', user: 'user'):
rabbitmq_1  | client unexpectedly closed TCP connection
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | accepting AMQP connection <0.673.0> (172.20.0.3:44670 -> 172.20.0.2:5672)
rabbitmq_1  | 
rabbitmq_1  | =INFO REPORT==== 5-Jun-2018::16:46:32 ===
rabbitmq_1  | connection <0.673.0> (172.20.0.3:44670 -> 172.20.0.2:5672): user 'user' authenticated and granted access to vhost 'test'
```

**Comment by maingoh** [comment]
I just tried to change the publish part : 

`connection.ensure` + `retry=False` => Not working: 
```python
send = connection.ensure(producer, producer.publish, errback=errback)
send({"hey": "you"}, routing_key=queue.name)
```

No `connection.ensure` + `retry=True` => Not working:
```python
producer.publish({"hey": "you"}, routing_key=queue.name, retry=True)
```

No `connection.ensure` + `retry=False` + manual retry => Working:
```python
import socket
from amqp.exceptions import RecoverableConnectionError

while True:
    try:
        producer.publish({"hey": "you"}, routing_key=queue.name)
        break
    except (socket.error, RecoverableConnectionError):
        logging.error(traceback.format_exc())
        producer.revive(connection.default_channel)
```

The last part seems to be the workaround @stuartspotlight use. But it probably doesn't handle reconnection well ..

**Comment by maingoh** [comment]
The bug seems to appears since kombu 4.0.0. It is working as expected in 3.0.37.

**Comment by maingoh** [comment]
With a similar code to the one I posted earlier and in heavy publish/consuming I got the exact same stacktrace as @stuartspotlight  : 
```python
python_client  |   File "./app/client.py", line 268, in send
python_client  |     self.ensure(self.producer, self.producer.publish, data, exchange=exchange, routing_key=routing_key, retry=True, retry_policy=retry_policy, correlation_id=correlation_id, **properties)
python_client  |   File "./app/client.py", line 161, in ensure
python_client  |     return f(*args, **kwargs)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 494, in _ensured
python_client  |     return fun(*args, **kwargs)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/kombu/messaging.py", line 181, in publish
python_client  |     exchange_name, declare,
python_client  |   File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 494, in _ensured
python_client  |     return fun(*args, **kwargs)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/kombu/messaging.py", line 203, in _publish
python_client  |     mandatory=mandatory, immediate=immediate,
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/channel.py", line 1743, in basic_publish_confirm
python_client  |     self.wait(spec.Basic.Ack)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/abstract_channel.py", line 79, in wait
python_client  |     self.connection.drain_events(timeout=timeout)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 491, in drain_events
python_client  |     while not self.blocking_read(timeout):
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 497, in blocking_read
python_client  |     return self.on_inbound_frame(frame)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/method_framing.py", line 55, in on_frame
python_client  |     callback(channel, method_sig, buf, None)
python_client  |   File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 500, in on_inbound_method
python_client  |     return self.channels[channel_id].dispatch_method(
python_client  | TypeError: 'NoneType' object has no attribute '__getitem__'
```

**Comment by stuartspotlight** [comment]
Sorry @maingoh, did you say you could fix the problem by removing the consuming section?

```
If I remove the consuming part, this doesn't disconnect in loop (just reconnect once and then publish as expected).
```
Does this mean you remove an explicit declaration of a consumer and use a consumer built into the "ConsumerProducerMixin" or that you don't use a consumer at all and only publish?

**Comment by maingoh** [comment]
Sorry, when i said "consuming part" i was meaning `queue.get`. I didn't try with a real Consumer but probably it won't disconnect with a consumer as it will keep the connection active ?

**Comment by stuartspotlight** [comment]
I'm using a  consumer and I'm having problems with the connection being closed on ack and not being able to recover. I tried the for loop trick and it just failed 30 times in a row. This is a major issue as its leading to messages being duplicated. When the ack fails I get this:-
```
finished running user code on Monster Manual.pdf at 2018-07-23 16:05:35.685752
RUNNING SEND TO NEXT PLUGIN
CREATING PAYLOAD STRING
SET UP THE PRODUCER
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
Trying to publish result to email-extraction
Output was published for Monster Manual.pdf
HAD TROUBLE ACKING TRYING ANOTHER 30 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 29 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 28 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 27 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 26 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 25 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 24 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 23 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 22 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 21 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 20 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 19 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 18 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 17 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 16 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 15 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 14 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 13 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 12 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 11 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 10 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 9 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 8 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 7 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 6 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 5 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 4 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 3 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 2 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
HAD TROUBLE ACKING TRYING ANOTHER 1 times
CONNECTION IS UP:- True
OUT CHANNEL IS UP:- True
---
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/nanowire_plugin/__init__.py", line 511, in send_to_next_plugin
    message.ack()
  File "/usr/local/lib/python3.6/site-packages/kombu/message.py", line 125, in ack
    self.channel.basic_ack(self.delivery_tag, multiple=multiple)
  File "/usr/local/lib/python3.6/site-packages/amqp/channel.py", line 1392, in basic_ack
    spec.Basic.Ack, argsig, (delivery_tag, multiple),
  File "/usr/local/lib/python3.6/site-packages/amqp/abstract_channel.py", line 47, in send_method
    raise RecoverableConnectionError('connection already closed')
amqp.exceptions.RecoverableConnectionError: connection already closed

==================================
COULD NOT ACK, KILLING MAIN THREAD

```

**Comment by auvipy** [comment]
this is a shame that the issue didn't get proper attention it needed.! sorry guys! did you found any workaround?

**Comment by stuartspotlight** [comment]
Manually repeating the publish command seems to work on the second attempt every time. The bigger issue is the failure to ack which is non-recoverable. This could be fixed by letting the user ack a message using the delivery tag and a different connection to the one it was received on but I'm not sure how I would go about implementing such a thing. I tried manually rebuilding a connection and assigning it to the message but the ack I sent didn't take, as in the ack raised no errors but rabbitmq did not remove it from the queue.

**Comment by maingoh** [comment]
It seems to me that when a connection occurs, it calls the transport collect which makes previous entities in a bad state ? 
If I patch the `connection.collect` by commenting the line `gc_transport(self._connection)`, both queues triggers a connection error in the first iteration, but then they get stable and don't trigger anything.

**Comment by jorgii** [comment]
I experienced the same issue. As a workaround it looks like adding `retry=True` fixes it.
```
producer.publish(
            body=message,
            exchange=exchange,
            declare=[queue],
            retry=True)
```

**Comment by emukans** [comment]
I experienced the same issue. The problem was in qos prefetch_size. The message queue was too large and prefetch was set to 0, it means that consumer tried to prefetch all queue. The solution would be to reduce qos prefetch_size or if it is not possible, purge message queue and resed it again in smaller batches.

**Comment by auvipy** [comment]
> I experienced the same issue. The problem was in qos prefetch_size. The message queue was too large and prefetch was set to 0, it means that consumer tried to prefetch all queue. The solution would be to reduce qos prefetch_size or if it is not possible, purge message queue and resed it again in smaller batches.

closing the issue based on this.
